src/v/storage/disk_log_impl.cc

```diff
@@ -2479,6 +2479,8 @@ ss::future<> disk_log_impl::do_truncate_prefix(truncate_prefix_config cfg) {
      * whose max offset falls below the new starting offset.
      */
     {
+        ssx::semaphore_units seg_rewrite_units
+          = co_await _segment_rewrite_lock.get_units();
```

Comment:
>I believe this can also be fixed by just taking a read/write lock on the
segment but the change is more intrusive as it requires additional
refactoring.

What's the gist of the refactor? If this is more intuitive and the change isn't big it might make sense?

Reply:
Routines for closing the segment acquire a write lock. The refactor I was thinking about is about refactoring these methods so that they can be passed the lock to avoid trying and failing to write lock it twice.